### PR TITLE
Switch debug registry mapping to `CodeInstance`

### PR DESCRIPTION
**Pull request by Keno**
Currently our debugging code maps from pointers to `MethodInstance`. However, I think it makes more sense to map to `CodeInstance` instead, since that's what we're actually compiling and as we're starting to make more use of external owners and custom specializations, we'll want to be able to annotate that in backtraces. This only adjusts the internal data structures - any actual printing changes for those sorts of use cases will have to come separately.

### CODE REVIEW

File: base/stacktraces.jl [hunk 1]
```diff
@@ -137,16 +137,26 @@ function lookup(ip::Base.InterpreterIP)
...
     else
+        func = top_level_scope_sym
+        file = empty_sym
+        line = Int32(0)
         if code isa Core.CodeInstance
             codeinfo = code.inferred::CodeInfo
+            def = code.def
+            if isa(def, Core.ABIOverride)
+                def = def.def
+            end
```
Comment (vtjnash) [inline base/stacktraces.jl:148]: The arbitrary proliferation of these checks seems a bad sign, but that is unrelated to this PR (which looks good to me)

**Review by vtjnash — APPROVED** [review]

File: base/stacktraces.jl (hunk 1 above)
Comment (Keno) [reply base/stacktraces.jl:148]: Yeah, we just need to put these behind a better accessor. I'll clean that up in the future.
